ui/src/app/devices/views/DeviceDetails/DeviceNetwork/EditInterface/EditInterface.test.tsx

```diff
@@ -0,0 +1,229 @@
+          systemId={systemId}
+        />
+      </Provider>
+    );
+    const wrapper = mount(<Proxy />);
+    const formValues = {
+      ip_address: "192.168.1.1",
+      ip_assignment: DeviceIpAssignment.STATIC,
+      mac_address: "11:22:33:44:55:66",
+      name: "eth123",
+      subnet: 2,
+      tags: ["tag1", "tag2"],
+    };
+    submitFormikForm(wrapper, formValues);
+    const updatingInterface = jest.spyOn(deviceSelectors, "getStatusForDevice");
+    updatingInterface.mockReturnValue(true);
+    // Make the component rerender with the new value.
+    store.dispatch({ type: "" });
+    wrapper.setProps({});
+    updatingInterface.mockReturnValue(false);
+    // Make the component rerender with the new value.
+    store.dispatch({ type: "" });
+    wrapper.setProps({});
+    expect(closeForm).toHaveBeenCalled();
+  });
+
+  it("does not close the form if there is an error when updating the interface", () => {
+    const closeForm = jest.fn();
+    state.device.errors = null;
+    const store = mockStore(state);
+    const Proxy = ({ systemId = "abc123" }) => (
+      <Provider store={store}>
+        <EditInterface
+          nicId={nic.id}
+          closeForm={closeForm}
+          systemId={systemId}
+        />
+      </Provider>
+    );
+    const wrapper = mount(<Proxy />);
+    const formValues = {
+      ip_address: "192.168.1.1",
+      ip_assignment: DeviceIpAssignment.STATIC,
+      mac_address: "11:22:33:44:55:66",
+      name: "eth123",
+      subnet: 2,
+      tags: ["tag1", "tag2"],
+    };
+    submitFormikForm(wrapper, formValues);
+    const errors = jest.spyOn(deviceSelectors, "eventErrorsForDevices");
+    errors.mockReturnValue([
+      deviceEventErrorFactory({
+        event: "updateInterface",
+      }),
+    ]);
+    const updatingInterface = jest.spyOn(deviceSelectors, "getStatusForDevice");
+    updatingInterface.mockReturnValue(true);
+    // Make the component rerender with the new value.
+    store.dispatch({ type: "" });
+    wrapper.setProps({});
+    updatingInterface.mockReturnValue(false);
+    // Make the component rerender with the new value.
+    store.dispatch({ type: "" });
+    wrapper.setProps({});
+    expect(closeForm).not.toHaveBeenCalled();
+  });
+
+  it("does not close the form if there is an error when submitting the form multiple times", () => {
+    const closeForm = jest.fn();
+    state.device.errors = null;
+    const store = mockStore(state);
+    const Proxy = ({ systemId = "abc123" }) => (
+      <Provider store={store}>
+        <EditInterface
+          nicId={nic.id}
+          closeForm={closeForm}
+          systemId={systemId}
+        />
+      </Provider>
+    );
+    const wrapper = mount(<Proxy />);
+    const formValues = {
+      ip_address: "192.168.1.1",
+      ip_assignment: DeviceIpAssignment.STATIC,
+      mac_address: "11:22:33:44:55:66",
+      name: "eth123",
+      subnet: 2,
+      tags: ["tag1", "tag2"],
+    };
+    submitFormikForm(wrapper, formValues);
```

Comment:
Did you consider testing the edit interface in the way the user would use it?

It would give more confidence the UI works as expected and have the advantage of providing a good overview of how the component is expected to behave at a glance, whilst this is arguably quite difficult to read.

We have no way of knowing if the form actually works correctly or if it's accessible other that testing it manually  - and there's no end-to-end test.

Correct me if I'm wrong, but looks like the form submission on the `EditInterface` component could be broken for the user and the test would still pass (you're calling the `onSubmit` handler directly, and not testing the actual UI anywhere).

Reply:
We could consider adding some integration tests for the forms in the future, but this is a unit test for the surface area that this component interacts with without needing to test the internals of the components it wraps. We've found this approach to be a reliable way to have confidence in the component and has a very low maintenance burden.

A small number of integration tests could help to raise that confidence but with the move to TypeScript it is hard to mis-use these components by e.g. passing the wrong type of data.